File: Component/BusiLib/src/main/java-person/com/zq/person/adapter/PhotoAdapter.java
package com.zq.person.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.common.log.MyLog;
import com.common.view.recyclerview.RecyclerOnItemClickListener;
import com.component.busilib.R;
import com.zq.person.holder.EmptyPhotoHolder;
import com.zq.person.holder.PhotoAddHolder;
import com.zq.person.holder.PhotoViewHolder;
import com.zq.person.model.PhotoModel;

import java.util.ArrayList;
import java.util.List;

public class PhotoAdapter extends RecyclerView.Adapter {
    public final static String TAG = "PhotoAdapter";

    public final static int TYPE_PERSON_CARD = 1; // dialogcarnd
    public final static int TYPE_PERSON_CENTER = 2; // 个人中心
    public final static int TYPE_OTHER_PERSON_CENTER = 3;//他人个人中心

    List<PhotoModel> mDataList = new ArrayList<>();
    RecyclerOnItemClickListener mListener;
    PhotoManageListener mPhotoManageListener;

    boolean mHasUpdate;
    int mType;

    private int PHOTO_ADD_TYPE = 0;     //上传
    private int PHOTO_ITEM_TYPE = 1;    //照片
    private int PHOTO_ITEM_EMPTY = 2;   //空页面

    public PhotoAdapter(RecyclerOnItemClickListener mListener, int type) {
        this.mListener = mListener;
        this.mType = type;
        if (type == TYPE_PERSON_CENTER) {
            mHasUpdate = true;
        } else {
            mHasUpdate = false;
        }
    }

    public void setPhotoManageListener(PhotoManageListener photoManageListener) {
        mPhotoManageListener = photoManageListener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == PHOTO_ITEM_TYPE) {
            View view;
            if (mType == TYPE_PERSON_CARD) {
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.photo_card_item_view_layout, parent, false);
            } else {
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.photo_item_view_layout, parent, false);
            }
            PhotoViewHolder viewHolder = new PhotoViewHolder(view, mListener);
            viewHolder.setPhotoManageListener(mPhotoManageListener);
            return viewHolder;
        } else if (viewType == PHOTO_ADD_TYPE) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.photo_add_view_layout, parent, false);
            PhotoAddHolder viewHolder = new PhotoAddHolder(view, mListener);
            return viewHolder;
        } else if (viewType == PHOTO_ITEM_EMPTY) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.photo_empty_layout, parent, false);
            EmptyPhotoHolder viewHolder = new EmptyPhotoHolder(view);
            return viewHolder;
        }
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (isContainEmpty()) {
            EmptyPhotoHolder emptyPhotoHolder = (EmptyPhotoHolder) holder;
            return;
        }

        if (mHasUpdate) {
            if (position == 0) {
                ((PhotoAddHolder) holder).bindData(position);
            } else {
                PhotoModel photoModel = mDataList.get(position - 1);
                ((PhotoViewHolder) holder).bindData(photoModel, position);
            }
        } else {
            ((PhotoViewHolder) holder).bindData(mDataList.get(position), position);
        }
    }


    @Override
    public int getItemViewType(int position) {
        if (isContainEmpty()) {
            return PHOTO_ITEM_EMPTY;
        }

        if (position == 0 && mHasUpdate) {
            return PHOTO_ADD_TYPE;
        }

        return PHOTO_ITEM_TYPE;
    }

    private boolean isContainEmpty() {
        if (mType == TYPE_OTHER_PERSON_CENTER) {
            if (mDataList != null && mDataList.size() > 0) {
                return false;
            } else {
                return true;
            }
        }
        return false;
    }

    @Override
    public int getItemCount() {
        if (mHasUpdate) {
            return mDataList.size() + 1;
        }

        if (isContainEmpty()) {
            // 空页面
            return 1;
        }

        return mDataList.size();
    }

    public List<PhotoModel> getDataList() {
        return mDataList;
    }

    /**
     * 列表中上传成功的item个数
     *
     * @return
     */
    public int getSuccessNum() {
        int success = 0;
        for (PhotoModel photoModel : mDataList) {
            if (photoModel.getStatus() == PhotoModel.STATUS_SUCCESS) {
                success++;
            }
        }
        return success;
    }

    public void setDataList(List<PhotoModel> dataList) {
        if (dataList != null) {
            // 把未成功的挑选出来
            List<PhotoModel> unSuccessList = new ArrayList<>();
            for (PhotoModel photoModel : mDataList) {
                if (photoModel.getStatus() != PhotoModel.STATUS_SUCCESS) {
                    unSuccessList.add(photoModel);
                }
            }
            mDataList.clear();
            mDataList.addAll(unSuccessList);
            mDataList.addAll(dataList);
            notifyDataSetChanged();
        }
    }

    /**
     * 头部插入某条数据
     *
     * @param data
     */
    public void insertFirst(PhotoModel data) {
        mDataList.add(0, data);
        if (mHasUpdate) {
            notifyItemInserted(1);
        } else {
            notifyItemInserted(0);
        }
    }

    /**
     * 尾部插入一堆数据
     *
     * @param list
     */
    public void insertLast(List<PhotoModel> list) {
        int origin = mDataList.size();
        mDataList.addAll(list);
        if (mHasUpdate) {
            notifyItemRangeInserted(origin + 1, mDataList.size() - origin + 1);
        } else {
            notifyItemRangeInserted(origin, mDataList.size() - origin);
        }
    }

    public void delete(PhotoModel photoModel) {
        for (int i = 0; i < mDataList.size(); i++) {
            PhotoModel m = mDataList.get(i);
            if (m.equals(photoModel)) {
                MyLog.d(TAG, "delete" + " find i=" + i);
                mDataList.remove(i);
                if (mHasUpdate) {
                    notifyItemRemoved(i + 1);
                } else {
                    notifyItemRemoved(i);
                }
                return;
            }
        }
    }

    public void update(PhotoModel photoModel) {
        MyLog.d(TAG, "update" + " photoModel=" + photoModel);
        for (int i = 0; i < mDataList.size(); i++) {
            PhotoModel m = mDataList.get(i);
            if (m.equals(photoModel)) {
                MyLog.d(TAG, "update" + " find i=" + i);
                m.setStatus(photoModel.getStatus());
                if (mHasUpdate) {
                    notifyItemChanged(i + 1);
                } else {
                    notifyItemChanged(i);
                }
                return;
            }
        }
    }

    public int getPostionOfItem(PhotoModel model) {
        for (int i = 0; i < mDataList.size(); i++) {
            PhotoModel m = mDataList.get(i);
            if (m.equals(model)) {
                return i;
            }
        }
        return 0;
    }

    public interface PhotoManageListener {
        void delete(PhotoModel model);

        void reupload(PhotoModel model);
    }
}

File: Component/BusiLib/src/main/java-person/com/component/person/photo/adapter/PhotoAdapter.kt
package com.component.person.photo.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.common.log.MyLog
import com.component.busilib.R
import com.component.person.photo.holder.EmptyPhotoHolder
import com.component.person.photo.holder.PhotoAddHolder
import com.component.person.photo.holder.PhotoViewHolder
import com.component.person.photo.model.PhotoModel

import java.util.ArrayList

class PhotoAdapter(internal var mType: Int) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    val TAG = "PhotoAdapter"

    internal var mDataList: ArrayList<PhotoModel>? = ArrayList()
    private var mHasUpdate: Boolean = false

    private val PHOTO_ADD_TYPE = 0     //上传
    private val PHOTO_ITEM_TYPE = 1    //照片
    private val PHOTO_ITEM_EMPTY = 2   //空页面

    var mOnClickPhotoListener: ((view: View, position: Int, model: PhotoModel?) -> Unit)? = null
    var mOnClickAddPhotoListener: (() -> Unit)? = null
    var mDeleteListener: ((model: PhotoModel?) -> Unit)? = null
    var mReUploadListener: ((model: PhotoModel?) -> Unit)? = null

    interface PhotoManageListener {
        fun delete(model: PhotoModel)

        fun reupload(model: PhotoModel)
    }

    private val isContainEmpty: Boolean
        get() = if (mType == TYPE_OTHER_PERSON_CENTER) {
            !(mDataList != null && mDataList!!.size > 0)
        } else false

    // 把未成功的挑选出来
    var dataList: ArrayList<PhotoModel>?
        get() = mDataList
        set(dataList) {
            if (dataList != null) {
                val unSuccessList = ArrayList<PhotoModel>()
                for (photoModel in mDataList!!) {
                    if (photoModel.status != PhotoModel.STATUS_SUCCESS) {
                        unSuccessList.add(photoModel)
                    }
                }
                mDataList!!.clear()
                mDataList!!.addAll(unSuccessList)
                mDataList!!.addAll(dataList)
                notifyDataSetChanged()
            }
        }

    /**
     * 列表中上传成功的item个数
     *
     * @return
     */
    val successNum: Int
        get() {
            var success = 0
            for (photoModel in mDataList!!) {
                if (photoModel.status == PhotoModel.STATUS_SUCCESS) {
                    success++
                }
            }
            return success
        }

    init {
        mHasUpdate = mType == TYPE_PERSON_CENTER
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            PHOTO_ADD_TYPE -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.photo_add_view_layout, parent, false)
                return PhotoAddHolder(view, mOnClickAddPhotoListener)
            }
            PHOTO_ITEM_EMPTY -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.photo_empty_layout, parent, false)
                return EmptyPhotoHolder(view)
            }
            else -> {
                val view: View = if (mType == TYPE_PERSON_CARD) {
                    LayoutInflater.from(parent.context).inflate(R.layout.photo_card_item_view_layout, parent, false)
                } else {
                    LayoutInflater.from(parent.context).inflate(R.layout.photo_item_view_layout, parent, false)
                }
                return PhotoViewHolder(view, mOnClickPhotoListener, mDeleteListener, mReUploadListener)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (isContainEmpty) {
            return
        }

        if (mHasUpdate) {
            if (position != 0) {
                val photoModel = mDataList!![position - 1]
                (holder as PhotoViewHolder).bindData(photoModel, position)
            }
        } else {
            (holder as PhotoViewHolder).bindData(mDataList!![position], position)
        }
    }


    override fun getItemViewType(position: Int): Int {
        if (isContainEmpty) {
            return PHOTO_ITEM_EMPTY
        }

        return if (position == 0 && mHasUpdate) {
            PHOTO_ADD_TYPE
        } else PHOTO_ITEM_TYPE

    }

    override fun getItemCount(): Int {
        if (mHasUpdate) {
            return mDataList!!.size + 1
        }

        return if (isContainEmpty) {
            // 空页面
            1
        } else mDataList!!.size

    }

    /**
     * 头部插入某条数据
     *
     * @param data
     */
    fun insertFirst(data: PhotoModel) {
        mDataList!!.add(0, data)
        if (mHasUpdate) {
            notifyItemInserted(1)
        } else {
            notifyItemInserted(0)
        }
    }

    /**
     * 尾部插入一堆数据
     *
     * @param list
     */
    fun insertLast(list: List<PhotoModel>) {
        val origin = mDataList!!.size
        mDataList!!.addAll(list)
        if (mHasUpdate) {
            notifyItemRangeInserted(origin + 1, mDataList!!.size - origin + 1)
        } else {
            notifyItemRangeInserted(origin, mDataList!!.size - origin)
        }
    }

    fun delete(photoModel: PhotoModel) {
        for (i in mDataList!!.indices) {
            val m = mDataList!![i]
            if (m == photoModel) {
                MyLog.d(TAG, "delete find i=$i")
                mDataList!!.removeAt(i)
                if (mHasUpdate) {
                    notifyItemRemoved(i + 1)
                } else {
                    notifyItemRemoved(i)
                }
                return
            }
        }
    }

    fun update(photoModel: PhotoModel) {
        MyLog.d(TAG, "update photoModel=$photoModel")
        for (i in mDataList!!.indices) {
            val m = mDataList!![i]
            if (m == photoModel) {
                MyLog.d(TAG, "update find i=$i")
                m.status = photoModel.status
                if (mHasUpdate) {
                    notifyItemChanged(i + 1)
                } else {
                    notifyItemChanged(i)
                }
                return
            }
        }
    }

    fun getPostionOfItem(model: PhotoModel?): Int {
        for (i in mDataList!!.indices) {
            val m = mDataList!![i]
            if (m == model) {
                return i
            }
        }
        return 0
    }


    companion object {

        val TYPE_PERSON_CARD = 1 // dialogcarnd
        val TYPE_PERSON_CENTER = 2 // 个人中心
        val TYPE_OTHER_PERSON_CENTER = 3//他人个人中心
    }
}
